### PR TITLE
add scopes.unparse_scopes, refine intersect_scopes

### PR DESCRIPTION
**Pull request by minrk**
and fix warning condition for intersection overlap

- only warn when there's a group only on one side and a user only on the other,
  otherwise there is no lost information to warn about (group and/or defined on both sides)
- correctly resolve servers as sub-scopes of user

<del>I made scopes.intersect_scopes public because I think it might be used in the single-user auth (separate PR), but that might change by the time I finish.</del> No need to make intersect public, I think. Reverted that part.

### CODE REVIEW

File: jupyterhub/scopes.py [hunk 1]
```diff
@@ -208,6 +227,19 @@ def parse_scopes(scope_list):
     return parsed_scopes
 
 
+def unparse_scopes(parsed_scopes):
```
Comment (IvanaH8) [inline jupyterhub/scopes.py:230]: Nice!

File: jupyterhub/scopes.py [hunk 2]
```diff
@@ -13,60 +15,77 @@ class Scope(Enum):
...
+                entity: set(parsed_scopes_a[base][entity])
+                & set(parsed_scopes_b[base][entity])
                 for entity in common_entities
             }
 
-    if warn:
-        app_log.warning(
-            "[!user=, !group=] or [!user=, !server=] combinations of filters present, intersection between not considered. May result in lower than intended permissions."
-        )
+            if 'server' in all_entities and 'user' in all_entities:
```
Comment (IvanaH8) [inline jupyterhub/scopes.py:70]: What if we have a `[read:users:servers!group=A, 'read:users:servers!server=abc/xyz]` where abc is member of group A? Technically that should be allowed, shouldn't it? Is it reasonable to raise a warning here as well for such situation?

**Comment by IvanaH8** [comment]
Great, thank you for jumping on this! Just one comment, otherwise perfect.

File: jupyterhub/tests/test_scopes.py [hunk 3]
```diff
@@ -760,3 +760,66 @@ async def test_resolve_token_permissions(
...
+            True,
+        ),
+        # and the reverse, just in case
+        (
+            set(["users!user=x"]),
+            set(["users!group=y"]),
+            set([]),
+            True,
+        ),
+        # resolves server under user, without warning
```
Comment (IvanaH8) [inline jupyterhub/tests/test_scopes.py:807]: Maybe adding another one where it resolves to an empty set? Probably not needed as the code does check properly for the username match.

File: jupyterhub/scopes.py (hunk 2 above)
Comment (minrk) [reply jupyterhub/scopes.py:70]: Yes, the group->server relationship is the same as group->user. Good catch!

File: jupyterhub/tests/test_scopes.py (hunk 3 above)
Comment (minrk) [reply jupyterhub/tests/test_scopes.py:815]: Added, thanks!

File: jupyterhub/scopes.py (hunk 2 above)
Comment (minrk) [reply jupyterhub/scopes.py:78]: added tests and warning conditions for this

**Review by IvanaH8 — APPROVED** [review]
Great!

**Comment by minrk** [comment]
Thanks for the review!

**Pull request merged**